src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -1229,12 +1229,12 @@ public void M3()
                 // (5,23): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
                 //     public async void M()
                 Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "M").WithLocation(5, 23),
-                // (16,17): error CS0610: Field or property cannot be of type 'ArgIterator'
+                // (16,16): error CS0306: The type 'ArgIterator' may not be used as a type argument
                 //         return (default(ArgIterator), default(ArgIterator));
-                Diagnostic(ErrorCode.ERR_FieldCantBeRefAny, "default(ArgIterator)").WithArguments("System.ArgIterator").WithLocation(16, 17),
-                // (16,39): error CS0610: Field or property cannot be of type 'ArgIterator'
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "(default(ArgIterator), default(ArgIterator))").WithArguments("System.ArgIterator").WithLocation(16, 16),
+                // (16,16): error CS0306: The type 'ArgIterator' may not be used as a type argument
                 //         return (default(ArgIterator), default(ArgIterator));
-                Diagnostic(ErrorCode.ERR_FieldCantBeRefAny, "default(ArgIterator)").WithArguments("System.ArgIterator").WithLocation(16, 39)
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "(default(ArgIterator), default(ArgIterator))").WithArguments("System.ArgIterator").WithLocation(16, 16)
```

Comment:
To solve this across the board, it would make sense to move the constraints checks down into `TupleTypeSymbol.Create` (the overload that has diagnostics parameters).

```C#
        internal static NamedTypeSymbol Create(
            Location locationOpt,
            ImmutableArray<TypeSymbol> elementTypes,
            ImmutableArray<Location> elementLocations,
            ImmutableArray<string> elementNames,
            CSharpCompilation compilation,
            CSharpSyntaxNode syntax = null,
            DiagnosticBag diagnostics = null
            )
        {
           [...]

            return Create(underlyingType, elementNames, locationOpt, elementLocations);
            // we should check constraints before returning and use the locations here to get specific errors reported
        }
```

---
In reply to: [91579246](https://github.com/dotnet/roslyn/pull/15641#discussion_r91579246) [](ancestors = 91579246)